org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/context/MultiLevelCleanupTest.java

```diff
@@ -0,0 +1,284 @@
+		try {
+			Method m = testClass.getDeclaredMethod("getStaticResourceChecker");
+			return (AbstractResourceChecker<?>) m.invoke(null);
+		} catch (Exception e) {
+			throw Exceptions.duck(e);
+		}
+	}
+
+	static AbstractResourceChecker<?> getGlobalResourceChecker(Class<? extends MultiLevelCleanupTest> testClass) {
+		try {
+			Method m = testClass.getDeclaredMethod("getGlobalResourceChecker");
+			return (AbstractResourceChecker<?>) m.invoke(null);
+		} catch (Exception e) {
+			throw Exceptions.duck(e);
+		}
+	}
+
+	void setupResource() {
+		getResourceChecker().setupResource(currentScope);
+	}
+
+	@SuppressWarnings("unchecked")
+	void assertSetup(Scope... fromScopes) {
+		getStaticResourceChecker(getClass()).assertSetup(softly, currentScope, fromScopes);
+		getResourceChecker().assertSetup(softly, currentScope, fromScopes);
+	}
+
+	@SuppressWarnings("unchecked")
+	void assertNotSetup(Scope... fromScopes) {
+		getStaticResourceChecker(getClass()).assertNotSetup(softly, currentScope, fromScopes);
+		getResourceChecker().assertNotSetup(softly, currentScope, fromScopes);
+	}
+
+	OSGiSoftAssertions softly;
+
+	@BeforeAll
+	static void beforeAll(TestInfo info) {
+		staticCurrentScope = BEFORE_ALL;
+		@SuppressWarnings("unchecked")
+		Class<? extends MultiLevelCleanupTest> testClass = (Class<? extends MultiLevelCleanupTest>) info
+			.getTestClass()
+			.get();
+
+		staticSoftly = new OSGiSoftAssertions();
+
+		staticAssertSetup(testClass, BEFORE_CLASS);
+		staticAssertNotSetup(testClass, BEFORE_ALL, BEFORE_EACH, INNER_TEST, PARAMETERIZED_TEST, NESTED_BEFORE_EACH,
+			NESTED_TEST, NESTED_AFTER_EACH, AFTER_EACH, AFTER_ALL);
+		staticSetupResource(testClass);
+		staticAssertSetup(testClass, BEFORE_CLASS, BEFORE_ALL);
+		staticAssertNotSetup(testClass, BEFORE_EACH, INNER_TEST, PARAMETERIZED_TEST, NESTED_BEFORE_EACH, NESTED_TEST,
+			NESTED_AFTER_EACH, AFTER_EACH, AFTER_ALL);
+	}
+
+	Scope currentScope;
+
+	@BeforeEach
+	void beforeEach() {
+		currentScope = BEFORE_EACH;
+		softly = new OSGiSoftAssertions();
+		assertSetup(BEFORE_CLASS, BEFORE_ALL);
+		assertNotSetup(BEFORE_EACH, INNER_TEST, PARAMETERIZED_TEST, NESTED_BEFORE_EACH, NESTED_TEST, NESTED_AFTER_EACH,
+			AFTER_EACH, AFTER_ALL);
+		setupResource();
+		assertSetup(BEFORE_CLASS, BEFORE_ALL, BEFORE_EACH);
+		assertNotSetup(INNER_TEST, PARAMETERIZED_TEST, NESTED_BEFORE_EACH, NESTED_TEST, NESTED_AFTER_EACH, AFTER_EACH,
+			AFTER_ALL);
+	}
+
+	@Test
+	void innerTest() {
+		currentScope = INNER_TEST;
+		assertSetup(BEFORE_CLASS, BEFORE_ALL, BEFORE_EACH);
+		assertNotSetup(INNER_TEST, PARAMETERIZED_TEST, NESTED_BEFORE_EACH, NESTED_TEST, NESTED_AFTER_EACH, AFTER_EACH,
+			AFTER_ALL);
+		setupResource();
+		assertSetup(BEFORE_CLASS, BEFORE_ALL, BEFORE_EACH, INNER_TEST);
+		assertNotSetup(PARAMETERIZED_TEST, NESTED_BEFORE_EACH, NESTED_TEST, NESTED_AFTER_EACH, AFTER_EACH, AFTER_ALL);
+	}
+
+	@ParameterizedTest
+	@ValueSource(ints = {
+		1, 2, 3
+	})
+	void parameterizedTest(int parameter) {
```

Comment:
where is this parameter used?

Reply:
...or is this some trick to make the test execute multiple times? If so shouldn't `@RepeatedTest(3)` be used?

Reply:
In the test display name, and probably also internally in the test descriptor unique ID. It's not actually used in the test anywhere, but you can't have a parameterized test without parameters and I wanted to make sure the tests covered that case.

Reply:
I have documented the unused parameter.